# Chapter 6. Recommending the right products

## 6.1 Collaborative filtering and product recommendation

### Product recommender system

A product recommender system is a system with the goal of predicting and compiling a list of items that a customer is likely to purchase

**Collaborative filtering:** Based on previous user behaviors, such as pages that they viewed, products that they purchased, or ratings that they have given to different items. The collaborative filtering approach then uses this data to find similarities between users or items, and recommends the most similar items or contents to the users. The basic assumption behind the collaborative filtering method is that those who have viewed or purchased similar contents or products in the past are likely to view or purchase similar kinds of contents or products in the future. Thus, based on this assumption, if one person purchased items A, B, and C and another person purchased items A, B, and D in the past, then the first person is likely to purchase item D and the other person is likely to purchase the item C, as they share lots of similarities between them

**Content-based filtering:** Produce a list of recommendations based on the characteristics of an item or a user. It typically looks at the keywords that describe the characteristics of an item. The basic assumption behind the content-based filtering method is that the users are likely to view or purchase items that are similar to those items that they have bought or viewed in the past. For example, if a user has listened to some songs in the past, then the content-based filtering method will recommend similar kinds of songs that share similar characteristics to those songs that the user has already listened to 

### Collaborative filtering

**User-to-item matrix:** individual users in the rows and individual items in the columns. The value in each cell represent whether the given user bought the given item or not

**Cosine similarity:** To compute similarities between users

**Approaches:**
1. User-based approach: collaborative filtering uses the similarities between users: User-to-item matrix
2. Item-based approach: collaborative filtering uses the similarities between items: Item-to-user matrix

## 6.2 Building a product recommendation algorithm with python

In [3]:
import pandas as pd
df = pd.read_excel('/Users/boyuan/Desktop/OneDrive/Hands on data science for martketing/Online Retail.xlsx', sheet_name = 'Online Retail')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [4]:
from pandas_profiling import ProfileReport
profile = ProfileReport(df)
profile.to_widgets()

In [5]:
df = df.loc[df['Quantity'] > 0]

### Data preparation

1. Handle NaN values in the dataset
2. Build a customer-to-item matrix

Without correct values in the CustomerID field, we cannot build a proper recommendation system, since the collaborative filtering algorithm depends on the historical item purchase data for individual customers

### Handling NaNs in the CustomerID field

In [6]:
df['CustomerID'].isna().sum()

133361

In [7]:
df.loc[df['CustomerID'].isna()].head()

InvoiceNo StockCode                      Description  Quantity  \
622     536414     22139                              NaN        56   
1443    536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444    536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445    536544     21786               POLKADOT RAIN HAT          4   
1446    536544     21787            RAIN PONCHO RETROSPOT         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
622  2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445 2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446 2010-12-01 14:32:00       1.66         NaN  United Kingdom

Exclude the missing CustomerIds 

In [8]:
df = df.dropna(subset = ['CustomerID'])

In [9]:
df.shape

(397924, 8)

### Building a customer-item matrix

In [10]:
customer_item_matrix = df.pivot_table(
    index = 'CustomerID',
    columns = 'StockCode',
    values = 'Quantity',
    aggfunc = 'sum'
)
customer_item_matrix

StockCode   10002  10080  10120  10125  10133  10135  11001  15030  15034  \
CustomerID                                                                  
12346.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12347.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12348.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12349.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12350.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...           ...    ...    ...    ...    ...    ...    ...    ...    ...   
18280.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18281.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18282.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18283.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18287.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

StockCode   15036  ...  90214V  90214W  90214Y  90214Z  BANK CHARGES  C2  DOT  \
CustomerID         ...                                                          
12346.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
12347.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
12348.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
12349.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
12350.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
...           ...  ...     ...     ...     ...     ...           ...  ..  ...   
18280.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
18281.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
18282.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
18283.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   
18287.0       NaN  ...     NaN     NaN     NaN     NaN           NaN NaN  NaN   

StockCode     M  PADS  POST  
CustomerID                   
12346.0     NaN   NaN   NaN  
12347.0     NaN   NaN   NaN  
12348.0     NaN   NaN   9.0  
12349.0     NaN   NaN   1.0  
12350.0     NaN   NaN   1.0  
...         ...   ...   ...  
18280.0     NaN   NaN   NaN  
18281.0     NaN   NaN   NaN  
18282.0     NaN   NaN   NaN  
18283.0     2.0   NaN   NaN  
18287.0     NaN   NaN   NaN  

[4339 rows x 3665 columns]

In [11]:
customer_item_matrix = customer_item_matrix.applymap(lambda x: 1 if x > 0 else 0)
customer_item_matrix

StockCode   10002  10080  10120  10125  10133  10135  11001  15030  15034  \
CustomerID                                                                  
12346.0         0      0      0      0      0      0      0      0      0   
12347.0         0      0      0      0      0      0      0      0      0   
12348.0         0      0      0      0      0      0      0      0      0   
12349.0         0      0      0      0      0      0      0      0      0   
12350.0         0      0      0      0      0      0      0      0      0   
...           ...    ...    ...    ...    ...    ...    ...    ...    ...   
18280.0         0      0      0      0      0      0      0      0      0   
18281.0         0      0      0      0      0      0      0      0      0   
18282.0         0      0      0      0      0      0      0      0      0   
18283.0         0      0      0      0      0      0      0      0      0   
18287.0         0      0      0      0      0      0      0      0      0   

StockCode   15036  ...  90214V  90214W  90214Y  90214Z  BANK CHARGES  C2  DOT  \
CustomerID         ...                                                          
12346.0         0  ...       0       0       0       0             0   0    0   
12347.0         0  ...       0       0       0       0             0   0    0   
12348.0         0  ...       0       0       0       0             0   0    0   
12349.0         0  ...       0       0       0       0             0   0    0   
12350.0         0  ...       0       0       0       0             0   0    0   
...           ...  ...     ...     ...     ...     ...           ...  ..  ...   
18280.0         0  ...       0       0       0       0             0   0    0   
18281.0         0  ...       0       0       0       0             0   0    0   
18282.0         0  ...       0       0       0       0             0   0    0   
18283.0         0  ...       0       0       0       0             0   0    0   
18287.0         0  ...       0       0       0       0             0   0    0   

StockCode   M  PADS  POST  
CustomerID                 
12346.0     0     0     0  
12347.0     0     0     0  
12348.0     0     0     1  
12349.0     0     0     1  
12350.0     0     0     1  
...        ..   ...   ...  
18280.0     0     0     0  
18281.0     0     0     0  
18282.0     0     0     0  
18283.0     1     0     0  
18287.0     0     0     0  

[4339 rows x 3665 columns]

### Collaborative filtering

In user-based approach, compute similarities between users based on their item purchase history

In item-based approach, compute similarities between items based on which items are often bought together with which other items

In [12]:
from sklearn.metrics.pairwise import cosine_similarity

### User-based collaborative filtering and recommendations

In [13]:
# User-to-user similarity matrix
user_user_sim_matrix = pd.DataFrame(
    cosine_similarity(customer_item_matrix)
)
user_user_sim_matrix

0         1         2         3         4         5     6         7     \
0      1.0  0.000000  0.000000  0.000000  0.000000  0.000000   0.0  0.000000   
1      0.0  1.000000  0.063022  0.046130  0.047795  0.038484   0.0  0.025876   
2      0.0  0.063022  1.000000  0.024953  0.051709  0.027756   0.0  0.027995   
3      0.0  0.046130  0.024953  1.000000  0.056773  0.137137   0.0  0.030737   
4      0.0  0.047795  0.051709  0.056773  1.000000  0.031575   0.0  0.000000   
...    ...       ...       ...       ...       ...       ...   ...       ...   
4334   0.0  0.062318  0.000000  0.000000  0.000000  0.000000   0.0  0.041523   
4335   0.0  0.000000  0.000000  0.000000  0.000000  0.000000   0.0  0.049629   
4336   0.0  0.113776  0.000000  0.067574  0.000000  0.037582   0.0  0.000000   
4337   0.0  0.109364  0.170905  0.137124  0.044866  0.080278   0.0  0.113354   
4338   0.0  0.012828  0.083269  0.030475  0.000000  0.033898   0.0  0.000000   

          8         9     ...  4329      4330      4331      4332      4333  \
0     0.000000  0.000000  ...   0.0  0.000000  0.000000  0.000000  0.000000   
1     0.136641  0.094742  ...   0.0  0.029709  0.052668  0.000000  0.032844   
2     0.118262  0.146427  ...   0.0  0.064282  0.113961  0.000000  0.000000   
3     0.032461  0.144692  ...   0.0  0.105868  0.000000  0.000000  0.039014   
4     0.000000  0.033315  ...   0.0  0.000000  0.000000  0.000000  0.000000   
...        ...       ...  ...   ...       ...       ...       ...       ...   
4334  0.000000  0.000000  ...   0.0  0.000000  0.000000  0.000000  0.105409   
4335  0.000000  0.000000  ...   0.0  0.000000  0.000000  0.000000  0.000000   
4336  0.160128  0.079305  ...   0.0  0.174078  0.000000  0.000000  0.000000   
4337  0.034204  0.093170  ...   0.0  0.037184  0.016480  0.043602  0.000000   
4338  0.108324  0.000000  ...   0.0  0.000000  0.104383  0.000000  0.043396   

          4334      4335      4336      4337      4338  
0     0.000000  0.000000  0.000000  0.000000  0.000000  
1     0.062318  0.000000  0.113776  0.109364  0.012828  
2     0.000000  0.000000  0.000000  0.170905  0.083269  
3     0.000000  0.000000  0.067574  0.137124  0.030475  
4     0.000000  0.000000  0.000000  0.044866  0.000000  
...        ...       ...       ...       ...       ...  
4334  1.000000  0.119523  0.000000  0.000000  0.000000  
4335  0.119523  1.000000  0.000000  0.046613  0.000000  
4336  0.000000  0.000000  1.000000  0.017800  0.000000  
4337  0.000000  0.046613  0.017800  1.000000  0.096334  
4338  0.000000  0.000000  0.000000  0.096334  1.000000  

[4339 rows x 4339 columns]

In [14]:
user_user_sim_matrix.columns = customer_item_matrix.index

user_user_sim_matrix['CustomerID'] = customer_item_matrix.index
user_user_sim_matrix = user_user_sim_matrix.set_index('CustomerID')
user_user_sim_matrix

CustomerID  12346.0   12347.0   12348.0   12349.0   12350.0   12352.0  \
CustomerID                                                              
12346.0         1.0  0.000000  0.000000  0.000000  0.000000  0.000000   
12347.0         0.0  1.000000  0.063022  0.046130  0.047795  0.038484   
12348.0         0.0  0.063022  1.000000  0.024953  0.051709  0.027756   
12349.0         0.0  0.046130  0.024953  1.000000  0.056773  0.137137   
12350.0         0.0  0.047795  0.051709  0.056773  1.000000  0.031575   
...             ...       ...       ...       ...       ...       ...   
18280.0         0.0  0.062318  0.000000  0.000000  0.000000  0.000000   
18281.0         0.0  0.000000  0.000000  0.000000  0.000000  0.000000   
18282.0         0.0  0.113776  0.000000  0.067574  0.000000  0.037582   
18283.0         0.0  0.109364  0.170905  0.137124  0.044866  0.080278   
18287.0         0.0  0.012828  0.083269  0.030475  0.000000  0.033898   

CustomerID  12353.0   12354.0   12355.0   12356.0  ...  18273.0   18274.0  \
CustomerID                                         ...                      
12346.0         0.0  0.000000  0.000000  0.000000  ...      0.0  0.000000   
12347.0         0.0  0.025876  0.136641  0.094742  ...      0.0  0.029709   
12348.0         0.0  0.027995  0.118262  0.146427  ...      0.0  0.064282   
12349.0         0.0  0.030737  0.032461  0.144692  ...      0.0  0.105868   
12350.0         0.0  0.000000  0.000000  0.033315  ...      0.0  0.000000   
...             ...       ...       ...       ...  ...      ...       ...   
18280.0         0.0  0.041523  0.000000  0.000000  ...      0.0  0.000000   
18281.0         0.0  0.049629  0.000000  0.000000  ...      0.0  0.000000   
18282.0         0.0  0.000000  0.160128  0.079305  ...      0.0  0.174078   
18283.0         0.0  0.113354  0.034204  0.093170  ...      0.0  0.037184   
18287.0         0.0  0.000000  0.108324  0.000000  ...      0.0  0.000000   

CustomerID   18276.0   18277.0   18278.0   18280.0   18281.0   18282.0  \
CustomerID                                                               
12346.0     0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
12347.0     0.052668  0.000000  0.032844  0.062318  0.000000  0.113776   
12348.0     0.113961  0.000000  0.000000  0.000000  0.000000  0.000000   
12349.0     0.000000  0.000000  0.039014  0.000000  0.000000  0.067574   
12350.0     0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
...              ...       ...       ...       ...       ...       ...   
18280.0     0.000000  0.000000  0.105409  1.000000  0.119523  0.000000   
18281.0     0.000000  0.000000  0.000000  0.119523  1.000000  0.000000   
18282.0     0.000000  0.000000  0.000000  0.000000  0.000000  1.000000   
18283.0     0.016480  0.043602  0.000000  0.000000  0.046613  0.017800   
18287.0     0.104383  0.000000  0.043396  0.000000  0.000000  0.000000   

CustomerID   18283.0   18287.0  
CustomerID                      
12346.0     0.000000  0.000000  
12347.0     0.109364  0.012828  
12348.0     0.170905  0.083269  
12349.0     0.137124  0.030475  
12350.0     0.044866  0.000000  
...              ...       ...  
18280.0     0.000000  0.000000  
18281.0     0.046613  0.000000  
18282.0     0.017800  0.000000  
18283.0     1.000000  0.096334  
18287.0     0.096334  1.000000  

[4339 rows x 4339 columns]

Tell which customers are similar to others, and which customers have bought similar items to others

In [15]:
user_user_sim_matrix.loc[12350.0].sort_values(ascending = False)

CustomerID
12350.0    1.000000
17935.0    0.183340
12414.0    0.181902
12652.0    0.175035
16692.0    0.171499
             ...   
15953.0    0.000000
15952.0    0.000000
15951.0    0.000000
15950.0    0.000000
12346.0    0.000000
Name: 12350.0, Length: 4339, dtype: float64

Pick user 12350 and 17935:
1. First, we need to identify the items that the customer 12350 and 17935 have a;ready bought 
2. Then, we are going to find the products that the target customer 17935 has not purchased, but customer 12350 has. Since these two customers have bought similar items in the past, we are going to assume that the target customer 17935 has a high chance of purchasing the items that he or she has not bought, but customer 12350 has bought 
3. Lastly, we are going to use this list of items and recommend them to the target customer 17935

In [16]:
items_bought_by_A = set(customer_item_matrix.loc[12350.0].iloc[
    customer_item_matrix.loc[12350.0].to_numpy().nonzero()
].index)
items_bought_by_A

{20615,
 20652,
 21171,
 21832,
 21864,
 21866,
 21908,
 21915,
 22348,
 22412,
 22551,
 22557,
 22620,
 '79066K',
 '79191C',
 '84086C',
 'POST'}

In [17]:
items_bought_by_B = set(customer_item_matrix.loc[17935.0].iloc[
    customer_item_matrix.loc[17935.0].to_numpy().nonzero()
].index)
items_bought_by_B

{20657,
 20659,
 20828,
 20856,
 21051,
 21866,
 21867,
 22208,
 22209,
 22210,
 22211,
 22449,
 22450,
 22551,
 22553,
 22557,
 22640,
 22659,
 22749,
 22752,
 22753,
 22754,
 22755,
 23290,
 23292,
 23309,
 '85099B',
 'POST'}

In [18]:
items_to_recommend_to_B = items_bought_by_A - items_bought_by_B
items_to_recommend_to_B

{20615,
 20652,
 21171,
 21832,
 21864,
 21908,
 21915,
 22348,
 22412,
 22620,
 '79066K',
 '79191C',
 '84086C'}

In [19]:
df.loc[
    df['StockCode'].isin(items_to_recommend_to_B), 
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode')

Description
StockCode                                 
21832                 CHOCOLATE CALCULATOR
21915               RED  HARMONICA IN BOX 
22620          4 TRADITIONAL SPINNING TOPS
79066K                      RETRO MOD TRAY
21864      UNION JACK FLAG PASSPORT COVER 
79191C         RETRO PLASTIC ELEPHANT TRAY
21908        CHOCOLATE THIS WAY METAL SIGN
20615         BLUE POLKADOT PASSPORT COVER
20652           BLUE POLKADOT LUGGAGE TAG 
22348          TEA BAG PLATE RED RETROSPOT
22412      METAL SIGN NEIGHBOURHOOD WITCH 
21171                 BATHROOM METAL SIGN 
84086C             PINK/PURPLE RETRO RADIO

There is one main disadvantage of using user-based collaborative filtering. Recommendations are based on the individual customer's purchase history. For new customers, we are not going to have enough data to compare these new customers against the others. In order to handle this problem, we can use item-based collaborative filtering. 

### Item-based collaborative filtering and recommendations

In [20]:
item_item_sim_matrix = pd.DataFrame(
    cosine_similarity(customer_item_matrix.T)
)
item_item_sim_matrix

0         1         2         3         4         5         6     \
0     1.000000  0.000000  0.094868  0.090351  0.062932  0.098907  0.095346   
1     0.000000  1.000000  0.000000  0.032774  0.045655  0.047836  0.000000   
2     0.094868  0.000000  1.000000  0.057143  0.059702  0.041703  0.060302   
3     0.090351  0.032774  0.057143  1.000000  0.042644  0.044682  0.043073   
4     0.062932  0.045655  0.059702  0.042644  1.000000  0.280097  0.045002   
...        ...       ...       ...       ...       ...       ...       ...   
3660  0.029361  0.000000  0.000000  0.000000  0.036955  0.019360  0.055989   
3661  0.000000  0.000000  0.000000  0.000000  0.000000  0.104257  0.150756   
3662  0.066915  0.016182  0.070535  0.070535  0.070185  0.066184  0.106335   
3663  0.000000  0.000000  0.000000  0.000000  0.049752  0.000000  0.000000   
3664  0.078217  0.000000  0.010993  0.070669  0.021877  0.034383  0.058004   

          7         8         9     ...      3655  3656      3657  3658  \
0     0.047673  0.075593  0.090815  ...  0.000000   0.0  0.000000   0.0   
1     0.000000  0.082261  0.049413  ...  0.000000   0.0  0.000000   0.0   
2     0.060302  0.095618  0.028718  ...  0.000000   0.0  0.000000   0.0   
3     0.000000  0.051224  0.030770  ...  0.000000   0.0  0.000000   0.0   
4     0.060003  0.071358  0.057152  ...  0.000000   0.0  0.000000   0.0   
...        ...       ...       ...  ...       ...   ...       ...   ...   
3660  0.000000  0.000000  0.039996  ...  0.000000   0.0  0.000000   0.0   
3661  0.000000  0.000000  0.000000  ...  0.000000   0.0  0.000000   0.0   
3662  0.063801  0.059013  0.086089  ...  0.049875   0.0  0.040723   0.0   
3663  0.000000  0.000000  0.000000  ...  0.000000   0.0  0.000000   0.0   
3664  0.016573  0.026278  0.051301  ...  0.038866   0.0  0.031734   0.0   

          3659      3660  3661      3662      3663      3664  
0     0.000000  0.029361   0.0  0.066915  0.000000  0.078217  
1     0.000000  0.000000   0.0  0.016182  0.000000  0.000000  
2     0.000000  0.000000   0.0  0.070535  0.000000  0.010993  
3     0.000000  0.000000   0.0  0.070535  0.000000  0.070669  
4     0.000000  0.036955   0.0  0.070185  0.049752  0.021877  
...        ...       ...   ...       ...       ...       ...  
3660  0.000000  1.000000   0.0  0.026196  0.000000  0.020413  
3661  0.000000  0.000000   1.0  0.000000  0.000000  0.000000  
3662  0.089220  0.026196   0.0  1.000000  0.000000  0.077539  
3663  0.000000  0.000000   0.0  0.000000  1.000000  0.000000  
3664  0.017381  0.020413   0.0  0.077539  0.000000  1.000000  

[3665 rows x 3665 columns]

In [21]:
item_item_sim_matrix.columns = customer_item_matrix.T.index
item_item_sim_matrix['StockCode'] = customer_item_matrix.T.index
item_item_sim_matrix = item_item_sim_matrix.set_index('StockCode')
item_item_sim_matrix

StockCode     10002     10080     10120     10125     10133     10135  \
StockCode                                                               
10002      1.000000  0.000000  0.094868  0.090351  0.062932  0.098907   
10080      0.000000  1.000000  0.000000  0.032774  0.045655  0.047836   
10120      0.094868  0.000000  1.000000  0.057143  0.059702  0.041703   
10125      0.090351  0.032774  0.057143  1.000000  0.042644  0.044682   
10133      0.062932  0.045655  0.059702  0.042644  1.000000  0.280097   
...             ...       ...       ...       ...       ...       ...   
C2         0.029361  0.000000  0.000000  0.000000  0.036955  0.019360   
DOT        0.000000  0.000000  0.000000  0.000000  0.000000  0.104257   
M          0.066915  0.016182  0.070535  0.070535  0.070185  0.066184   
PADS       0.000000  0.000000  0.000000  0.000000  0.049752  0.000000   
POST       0.078217  0.000000  0.010993  0.070669  0.021877  0.034383   

StockCode     11001     15030     15034     15036  ...    90214V  90214W  \
StockCode                                          ...                     
10002      0.095346  0.047673  0.075593  0.090815  ...  0.000000     0.0   
10080      0.000000  0.000000  0.082261  0.049413  ...  0.000000     0.0   
10120      0.060302  0.060302  0.095618  0.028718  ...  0.000000     0.0   
10125      0.043073  0.000000  0.051224  0.030770  ...  0.000000     0.0   
10133      0.045002  0.060003  0.071358  0.057152  ...  0.000000     0.0   
...             ...       ...       ...       ...  ...       ...     ...   
C2         0.055989  0.000000  0.000000  0.039996  ...  0.000000     0.0   
DOT        0.150756  0.000000  0.000000  0.000000  ...  0.000000     0.0   
M          0.106335  0.063801  0.059013  0.086089  ...  0.049875     0.0   
PADS       0.000000  0.000000  0.000000  0.000000  ...  0.000000     0.0   
POST       0.058004  0.016573  0.026278  0.051301  ...  0.038866     0.0   

StockCode    90214Y  90214Z  BANK CHARGES        C2  DOT         M      PADS  \
StockCode                                                                      
10002      0.000000     0.0      0.000000  0.029361  0.0  0.066915  0.000000   
10080      0.000000     0.0      0.000000  0.000000  0.0  0.016182  0.000000   
10120      0.000000     0.0      0.000000  0.000000  0.0  0.070535  0.000000   
10125      0.000000     0.0      0.000000  0.000000  0.0  0.070535  0.000000   
10133      0.000000     0.0      0.000000  0.036955  0.0  0.070185  0.049752   
...             ...     ...           ...       ...  ...       ...       ...   
C2         0.000000     0.0      0.000000  1.000000  0.0  0.026196  0.000000   
DOT        0.000000     0.0      0.000000  0.000000  1.0  0.000000  0.000000   
M          0.040723     0.0      0.089220  0.026196  0.0  1.000000  0.000000   
PADS       0.000000     0.0      0.000000  0.000000  0.0  0.000000  1.000000   
POST       0.031734     0.0      0.017381  0.020413  0.0  0.077539  0.000000   

StockCode      POST  
StockCode            
10002      0.078217  
10080      0.000000  
10120      0.010993  
10125      0.070669  
10133      0.021877  
...             ...  
C2         0.020413  
DOT        0.000000  
M          0.077539  
PADS       0.000000  
POST       1.000000  

[3665 rows x 3665 columns]

1. First, for the given product that the target customer bought, we are going yo find the most similar items from the item-to-item similarity matrix.
2. Then, we are going to recommend these similar items to the customer, since those similar items were bought by other customers who have bought the product that the target customer initially bought. 

Assume a new customer just bought a product with StockCode 23166, and we want to include some products that this customer is the most likely to purchase in our marketing emails. The first thing we need to do is find the most similar items to the one with StockCode 23166

In [22]:
top_10_similar_items = list(
    item_item_sim_matrix
    .loc[23166]
    .sort_values(ascending = False)
    .iloc[:10]
    .index
)
top_10_similar_items

[23166, 23165, 23167, 22993, 23307, 22722, 22720, 22666, 23243, 22961]

In [23]:
df.loc[
    df['StockCode'].isin(top_10_similar_items),
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode').loc[top_10_similar_items]

Description
StockCode                                     
23166           MEDIUM CERAMIC TOP STORAGE JAR
23165            LARGE CERAMIC TOP STORAGE JAR
23167           SMALL CERAMIC TOP STORAGE JAR 
22993             SET OF 4 PANTRY JELLY MOULDS
23307      SET OF 60 PANTRY DESIGN CAKE CASES 
22722        SET OF 6 SPICE TINS PANTRY DESIGN
22720        SET OF 3 CAKE TINS PANTRY DESIGN 
22666          RECIPE BOX PANTRY YELLOW DESIGN
23243      SET OF TEA COFFEE SUGAR TINS PANTRY
22961                   JAM MAKING SET PRINTED

Personalizing the marketing messages with targeted product recommendations typically yields higher conversion rates from customers. Using an item-based collaborative filtering algorithm, you can now easily do product recommendations for both new and existing customers

## 6.3 Building a product recommendation algorithm with r